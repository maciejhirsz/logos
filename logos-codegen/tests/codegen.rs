use std::{error::Error, io, path::PathBuf};

#[rstest::rstest]
#[case("simple")]
#[case("no_error_lut")]
pub fn test_codegen(#[case] fixture: &str) -> Result<(), Box<dyn Error>> {
    let mut fixture_dir = PathBuf::new();
    fixture_dir.push(env!("CARGO_MANIFEST_DIR"));
    fixture_dir.push("tests");
    fixture_dir.push("data");
    fixture_dir.push(fixture);

    let input = fixture_dir.join("input.rs");
    fixture_dir.push("output.rs");
    let output_file = fixture_dir;

    let input = std::fs::read_to_string(input)?;
    let output = std::fs::read_to_string(&output_file)?;

    let generated = logos_codegen::generate(input.parse()?);
    let generated = generated.to_string();

    if std::env::var("BLESS_CODEGEN").is_ok_and(|value| value == "1") {
        std::fs::write(&output_file, &generated)?;
        return Ok(());
    }

    assert_eq!(generated, output, "Codegen test failed: `{fixture}`, run tests again with env var `BLESS_CODEGEN=1` to bless these changes");

    Ok(())
}
