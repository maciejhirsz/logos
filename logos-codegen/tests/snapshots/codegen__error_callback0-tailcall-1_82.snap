---
source: logos-codegen/tests/codegen.rs
expression: formatted
---
impl<'s> ::logos::Logos<'s> for TokenA {
    type Error = &'static str;
    type Extras = Vec<&'static str>;
    type Source = str;
    fn lex(
        lex: &mut ::logos::Lexer<'s, Self>,
    ) -> std::option::Option<std::result::Result<Self, Self::Error>> {
        use ::logos::internal::{
            LexerInternal, CallbackRetVal, CallbackResult, SkipRetVal, SkipResult,
        };
        use std::result::Result as _Result;
        use std::option::Option as _Option;
        use ::logos::Logos;
        type _Lexer<'s> = ::logos::Lexer<'s, TokenA>;
        #[inline]
        fn _make_error<'s>(lex: &mut _Lexer<'s>) -> <TokenA as Logos<'s>>::Error {
            let lex = lex;
            let error = {
                lex.extras.push("a");
                "a"
            };
            error.into()
        }
        fn state0<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<TokenA, <TokenA as Logos<'s>>::Error>> {
            match lex.read::<u8>(offset) {
                None if lex.offset() == offset => return None,
                other => {
                    lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
                    return Some(Err(_make_error(lex)));
                }
            }
        }
        state0(lex, lex.offset())
    }
}
